pkg/ring/kv/consul/client.go

```diff
@@ -243,31 +249,38 @@ func (c *Client) WatchPrefix(ctx context.Context, prefix string, f func(string,
 		}
 
 		kvps, meta, err := c.kv.List(prefix, queryOptions.WithContext(ctx))
-		if err != nil || kvps == nil {
+		// kvps being nil here is not an error -- quite the opposite. Consul returns index,
+		// which makes next query blocking, so there is no need to detect this and act on it.
+		if err != nil {
 			level.Error(util.Logger).Log("msg", "error getting path", "prefix", prefix, "err", err)
 			backoff.Wait()
 			continue
 		}
 		backoff.Reset()
 
-		skip := false
-		index, skip = checkLastIndex(index, meta.LastIndex)
+		newIndex, skip := checkLastIndex(index, meta.LastIndex)
 		if skip {
 			continue
 		}
 
 		for _, kvp := range kvps {
+			// We asked for values newer than 'index', but Consul returns all values below given prefix,
+			// even those that haven't changed. We don't need to report all of them as updated.
+			if index > 0 && kvp.ModifyIndex <= index && kvp.CreateIndex <= index {
+				continue
+			}
+
 			out, err := c.codec.Decode(kvp.Value)
 			if err != nil {
 				level.Error(util.Logger).Log("msg", "error decoding list of values for prefix:key", "prefix", prefix, "key", kvp.Key, "err", err)
 				continue
 			}
-			// We should strip the prefix from the front of the key.
-			key := strings.TrimPrefix(kvp.Key, prefix)
```

Comment:
Removing this, isn't a breaking change?

Reply:
Yes, but it also brings consistency between Consul and Etcd clients. We (I) need to find out if it was relied on. According to @cstyan: "I think that was for ha tracking for something". Fortunately, this method is only used in HA tracking code, so hopefully it shouldn't be that hard.

Reply:
Right. It's used only by the HA tracker, and it watches the prefix `""`, so the change should be harmless 👍

Reply:
You're right. Thanks for checking!